design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,373 @@
+| 2  | Policy is loaded w/o "provider-uri" variable or value is missing                          | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (RoleNotAuthorizedOnWebservice) is returned and logged | Integration               |            |
+| 3  | Policy is loaded w/o defining authenticator webservice                                    | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (WebserviceNotFound) is returned and logged            | Integration               |            |
+| 4  | An incorrectly/undefined provider-uri (Conjur couldn't connect to provider)               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (ProviderDiscoveryTimeout) is returned and logged      | Integration               |            |
+| 5  | Conjur can make contact with provider but couldn't confirm token signature                | VM sends authentication request to Conjur with invalid Azure token  | 500 Internal Server Error response with an error body with the relevant error message (ProviderTokenInvalid) is returned and logged          | Integration               |            |
+| 6  | One or more annotations are missing from the host policy **                               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (MissingRequestParam) is returned and logged           | Integration               |            |
+| 7  | One or more fields in Azure token is missing                                              | VM sends authentication request to Conjur with invalid Azure token  | 500 Internal Server Error response with an error body with the relevant error message (TokenFieldNotFoundOrEmpty) is returned and logged     | Integration               |            |
+| 8  | One or more fields in the Azure token don't match the Conjur host annotations             | VM sends authentication request to Conjur with valid   Azure token  | 500 Internal Server Error response with an error body with the relevant error message (InvalidApplicationIdentity) is returned and logged    | Integration               |            |
+
+### Unit tests
+
+1. Validate that `validate_azure_token` properly decodes the of JWT token
+
+1. Validate that `validate_application_identity` accurately compares the Conjur host identity and Azure token correctly
+
+1. If `validate_azure_token` or `validate_application_identity` raise an error then it propagates to calling `Authentication::Authenticate` class
```

Comment:
So this is a UT on `Authentication::Authenticate`? This class doesn't know `validate_azure_token` or `validate_application_identity` so it's not a valid UT.

Reply:
From my understanding, `Authentication::Authenticate` runs the general authenticator validations and part of this sequence of validations includes `validate_credentials` which calls `valid?` for the relevant authenticator (authn-azure). This function triggers the authenticator-specific validations. 

So my thought was that we need to ensure that when one of the authenticator-specific checks fails, it is propagated. 

Is the act of propagation an integration test?

Reply:
this test is already implemented here: https://github.com/cyberark/conjur/blob/master/spec/app/domain/authentication/authenticate_spec.rb